File: digiposte_api.c
#include <json-c/json.h>
#include "digiposte_api.h"

static int read_fd, write_fd;

int init_api(const char *authorization)
{
    int cts_pipe[2], stc_pipe[2], child;
    char args_read_fd[4], args_write_fd[4];
    char buf[6];
    
    if (pipe(cts_pipe) == -1 || pipe(stc_pipe) == -1) {
        perror("pipe()");
        return -1;
    }
    
    child = fork();
    if (child == -1) {
        perror("fork()");
        return -1;
    }
    else if (child == 0) {
        //AppArmor change profile here
        
        close(stc_pipe[0]);
        close(cts_pipe[1]);
        
        snprintf(args_read_fd, 4, "%d", cts_pipe[0]);
        snprintf(args_write_fd, 4, "%d", stc_pipe[1]);
        
        execlp("python3", "python3", DGP_API_SUBSYSTEM, "--server", args_read_fd, args_write_fd, "--token", authorization, NULL);
        perror("execlp()");
        exit(-errno);
    }
    
    close(stc_pipe[1]);
    close(cts_pipe[0]);
    
    read_fd = stc_pipe[0];
    write_fd = cts_pipe[1];
    
    if (read(read_fd, buf, 6) == -1) {
        perror("read()");
        return -1;
    }
    
    return 0;
}

void free_api()
{
    close(read_fd);
    close(write_fd);
}

static void construct_folder_rec(c_folder *folder, json_object *root)
{
    int n, i;
    json_object *j_folders, *tmp, *field_id, *field_name, *field_location;
    c_folder *new;

    j_folders = json_object_object_get(root, "folders");
    n = json_object_array_length(j_folders);
    for (i=0; i<n; i++) {
        tmp = json_object_array_get_idx(j_folders, i);
        field_id = json_object_object_get(tmp, "id");
        field_name = json_object_object_get(tmp, "name");
        field_location = json_object_object_get(tmp, "location");
        if (strncmp(json_object_get_string(field_location), "TRASH", 6) != 0)
            new = add_folder(folder, json_object_get_string(field_id), json_object_get_string(field_name));

        construct_folder_rec(new, tmp);
    }
}

c_folder* get_folders()
{
    c_folder *folder;
    json_object *root;
    resp_stuct *rs;
    char *tmp;
    int r;

    rs = malloc(sizeof(resp_stuct));
    if (rs == NULL) {
        perror("malloc()");
        return NULL;
    }
    rs->ptr = malloc(BUF_SIZE*sizeof(char));
    if (rs->ptr == NULL) {
        perror("malloc()");
        free(rs);
        return NULL;
    }
    rs->response_actual_size = 0;
    rs->response_allocated_size = BUF_SIZE;
    
    r = write(write_fd, "get_folders_tree\n", 17);
    if (r != 17) {
        perror("write()");
        free(rs->ptr);
        free(rs);
        return NULL;
    }
    
    do {
        r = read(read_fd, rs->ptr + rs->response_actual_size, CHUNK_SIZE);
        if (r == -1) {
            perror("read()");
            free(rs->ptr);
            free(rs);
            return NULL;
        }
        
        rs->response_actual_size += r;
        if (rs->ptr[rs->response_actual_size-1] == '\0') break;
        
        if (r == CHUNK_SIZE && rs->response_allocated_size <= rs->response_actual_size + CHUNK_SIZE) {
            tmp = realloc(rs->ptr, rs->response_actual_size + CHUNK_SIZE*2);
            if (tmp == NULL) {
                perror("realloc()");
                free(rs->ptr);
                free(rs);
                return NULL;
            }
            rs->ptr = tmp;
            rs->response_allocated_size = rs->response_actual_size + CHUNK_SIZE*2;
        }
    } while (r == CHUNK_SIZE);
    
    if (rs->ptr[0] == 'e' && rs->ptr[1] == 'r' && rs->ptr[2] == 'r') {
        fputs("API returned an error\n", stderr);
        free(rs->ptr);
        free(rs);
        return NULL;
    }

    root = json_tokener_parse(rs->ptr);
    if (root == NULL) {
        fputs("json_tokener_parse(): error\n", stderr);
        free(rs->ptr);
        free(rs);
        return NULL;
    }

    folder = add_folder(NULL, "root-000000000000000000000000000", NULL);
    if (folder == NULL) {
        free(rs->ptr);
        free(rs);
        return NULL;
    }

    construct_folder_rec(folder, root);

    json_object_put(root);
    free(rs->ptr);
    free(rs);

    return folder;
}

int get_folder_content(c_folder *folder)
{
    json_object *root, *j_file, *field_id, *field_name, *field_size, *tmp;
    resp_stuct *rs;
    char req[64], *tmp_ptr;
    int r, i, n;

    rs = malloc(sizeof(resp_stuct));
    if (rs == NULL) {
        perror("malloc()");
        return -1;
    }
    rs->ptr = malloc(BUF_SIZE*sizeof(char));
    if (rs->ptr == NULL) {
        perror("malloc()");
        free(rs);
        return -1;
    }
    rs->response_actual_size = 0;
    rs->response_allocated_size = BUF_SIZE;
    
    if (folder->id[0] == 'r') {
        memcpy(req, "get_folder_content", 19);
        req[19] = '\n';
        i = 20;
    }
    else {
        memcpy(req, "get_folder_content", 19);
        memcpy(req+19, folder->id, 32);
        req[51] = '\n';
        i = 52;
    }
    
    r = write(write_fd, req, i);
    if (r != i) {
        perror("write()");
        free(rs->ptr);
        free(rs);
        return -1;
    }
    
    do {
        r = read(read_fd, rs->ptr + rs->response_actual_size, CHUNK_SIZE);
        if (r == -1) {
            perror("read()");
            free(rs->ptr);
            free(rs);
            return -1;
        }
        
        rs->response_actual_size += r;
        if (rs->ptr[rs->response_actual_size-1] == '\0') break;
        
        if (r == CHUNK_SIZE && rs->response_allocated_size <= rs->response_actual_size + CHUNK_SIZE) {
            tmp_ptr = realloc(rs->ptr, rs->response_actual_size + CHUNK_SIZE*2);
            if (tmp_ptr == NULL) {
                perror("realloc()");
                free(rs->ptr);
                free(rs);
                return -1;
            }
            rs->ptr = tmp_ptr;
            rs->response_allocated_size = rs->response_actual_size + CHUNK_SIZE*2;
        }
    } while (r == CHUNK_SIZE);
    
    if (rs->ptr[0] == 'e' && rs->ptr[1] == 'r' && rs->ptr[2] == 'r') {
        fputs("API returned an error\n", stderr);
        free(rs->ptr);
        free(rs);
        return -1;
    }

    root = json_tokener_parse(rs->ptr);
    if (root == NULL) {
        fputs("json_tokener_parse(): error\n", stderr);
        free(rs->ptr);
        free(rs);
        return -1;
    }

    j_file = json_object_object_get(root, "documents");
    n = json_object_array_length(j_file);
    for (i=0; i<n; i++) {
        tmp = json_object_array_get_idx(j_file, i);
        field_id = json_object_object_get(tmp, "id");
        field_name = json_object_object_get(tmp, "filename");
        field_size = json_object_object_get(tmp, "size");
        add_file(folder, json_object_get_string(field_id), json_object_get_string(field_name), json_object_get_int(field_size));
    }

    folder->files_loaded = 1;

    json_object_put(root);
    free(rs->ptr);
    free(rs);

    return 0;
}

int get_file(const c_file *file, const char *dest_path)
{
    char req[128], resp[4];
    int r, len;
    
    memcpy(req, "get_file", 9);
    memcpy(req+9, file->id, 32);
    req[41] = '\0';
    len = strlen(dest_path);
    memcpy(req+42, dest_path, len);
    req[42+len] = '\n';
    
    r = write(write_fd, req, len+43);
    if (r != len+43) {
        perror("write()");
        return -1;
    }
    
    r = read(read_fd, resp, 4);
    if (r == -1) {
        perror("read()");
        return -1;
    }
    
    if (resp[0] == 'e' && resp[1] == 'r' && resp[2] == 'r') {
        fputs("API returned an error\n", stderr);
        return -1;
    }
    if (resp[0] != 'O' || resp[1] != 'K') {
        fputs("API returned an unexpected error\n", stderr);
        return -1;
    }

    return 0;
}

int create_folder(const char *name, const char *parent_id, char *new_id)
{
    char req[128], resp[33];
    int r, i, name_len;
    
    memcpy(req, "create_folder", 14);
    name_len = strlen(name);
    memcpy(req+14, name, name_len+1);
    if (parent_id[0] == 'r') {
        i = 15;
    }
    else {
        memcpy(req+15+name_len, parent_id, 32);
        i = 47;
    }
    req[name_len+i] = '\n';
    
    r = write(write_fd, req, name_len+i+1);
    if (r != name_len+i+1) {
        perror("write()");
        return -1;
    }
    
    r = read(read_fd, resp, 33);
    if (r == -1) {
        perror("read()");
        return -1;
    }
    
    if (resp[0] == 'e' && resp[1] == 'r' && resp[2] == 'r') {
        fputs("API returned an error\n", stderr);
        return -1;
    }
    
    memcpy(new_id, resp, 32);

    return 0;
}

int rename_object(const char *id, const char *new_name, const char is_file)
{
    int r, name_len;
    char req[128], resp[4];
    
    memcpy(req, "rename_object", 14);
    if (is_file) req[14] = '1';
    else req[14] = '0';
    req[15] = '\0';
    memcpy(req+16, id, 32);
    req[48] = '\0';
    name_len = strlen(new_name);
    memcpy(req+49, new_name, name_len);
    req[name_len+49] = '\n';
    
    r = write(write_fd, req, name_len+50);
    if (r != name_len+50) {
        perror("write()");
        return -1;
    }
    
    r = read(read_fd, resp, 4);
    if (r == -1) {
        perror("read()");
        return -1;
    }
    
    if (resp[0] == 'e' && resp[1] == 'r' && resp[2] == 'r') {
        fputs("API returned an error\n", stderr);
        return -1;
    }
    if (resp[0] != 'O' || resp[1] != 'K') {
        fputs("API returned an unexpected error\n", stderr);
        return -1;
    }

    return 0;
}

int delete_object(const char *id, const char is_file)
{
    int r;
    char req[128], resp[4];
    
    memcpy(req, "delete_object", 14);
    if (is_file) req[14] = '1';
    else req[14] = '0';
    req[15] = '\0';
    memcpy(req+16, id, 32);
    req[48] = '\n';
    
    r = write(write_fd, req, 49);
    if (r != 49) {
        perror("write()");
        return -1;
    }

    r = read(read_fd, resp, 4);
    if (r == -1) {
        perror("read()");
        return -1;
    }
    
    if (resp[0] == 'e' && resp[1] == 'r' && resp[2] == 'r') {
        fputs("API returned an error\n", stderr);
        return -1;
    }
    if (resp[0] != 'O' || resp[1] != 'K') {
        fputs("API returned an unexpected error\n", stderr);
        return -1;
    }

    return 0;
}

int move_object(const char *id, const char *to_folder_id, const char is_file)
{
    int r, i;
    char req[128], resp[4];
    
    memcpy(req, "move_object", 12);
    if (is_file) req[12] = '1';
    else req[12] = '0';
    req[13] = '\0';
    memcpy(req+14, id, 32);
    req[46] = '\0';
    if (to_folder_id[0] == 'r') {
        req[47] = '\n';
        i = 48;
    }
    else {
        memcpy(req+47, to_folder_id, 32);
        req[79] = '\n';
        i = 80;
    }
    
    r = write(write_fd, req, i);
    if (r != i) {
        perror("write()");
        return -1;
    }
    
    r = read(read_fd, resp, 4);
    if (r == -1) {
        perror("read()");
        return -1;
    }
    
    if (resp[0] == 'e' && resp[1] == 'r' && resp[2] == 'r') {
        fputs("API returned an error\n", stderr);
        return -1;
    }
    if (resp[0] != 'O' || resp[1] != 'K') {
        fputs("API returned an unexpected error\n", stderr);
        return -1;
    }

    return 0;
}

int upload_file(const c_file *file, const char *to_folder_id, char *new_id)
{
    int r, len, i;
    char req[128], resp[33];
    
    memcpy(req, "upload_file", 12);
    if (to_folder_id[0] == 'r') {
        req[12] = '\0';
        i = 13;
    }
    else {
        memcpy(req+12, to_folder_id, 32);
        req[44] = '\0';
        i = 45;
    }
    len = strlen(file->cache_path);
    memcpy(req+i, file->cache_path, len+1);
    i += len+1;
    len = strlen(file->name);
    memcpy(req+i, file->name, len+1);
    i += len+1;
    snprintf(req+i, 10, "%ld\n", file->size);
    i += strlen(req+i);
    
    r = write(write_fd, req, i);
    if (r != i) {
        perror("write()");
        return -1;
    }
    
    r = read(read_fd, resp, 33);
    if (r == -1) {
        perror("read()");
        return -1;
    }
    
    if (resp[0] == 'e' && resp[1] == 'r' && resp[2] == 'r') {
        fputs("API returned an error\n", stderr);
        return -1;
    }
    
    memcpy(new_id, resp, 32);

    return 0;
}

File: DigiposteAPI.py
#!/usr/bin/python3

import os
import sys
import argparse
import requests
import webview
import threading
import time

class AuthAPI:
    def __init__(self):
        self.token = None

    def receive_token(self, token):
        self.token = token
        return "OK"

class DigiposteAPI():
    def __init__(self, token=None):
        self._session = requests.Session()
        
        if token is None:
            self._token = self._authenticate()
            if self._token is None:
                print("No authentication token retrieved")
                exit(1)
        else:
            self._token = token[0]
        
        self._session.headers.update({"Authorization": "Bearer {}".format(self._token), "Accept": "application/json"})
    
    def _authenticate(self):
        def inject_js(window):
            js = f"""
            (function() {{
                try {{
                    const token = sessionStorage.getItem("access_token");
                    if (token) {{
                        window.pywebview.api.receive_token(token);
                    }}
                }} catch (e) {{
                    console.log("Erreur JS:", e);
                }}
            }})();
            """
            window.evaluate_js(js)
        
        api = AuthAPI()

        window = webview.create_window(
            "Connexion Digiposte",
            "https://secure.digiposte.fr/identification-plus",
            js_api=api,
            width=480,
            height=720
        )

        def monitor():
            while api.token is None:
                time.sleep(1)
                inject_js(window)

            window.destroy()

        threading.Thread(target=monitor, daemon=True).start()
        webview.start()
        
        return api.token
    
    def disconnect(self):
        pass
    
    def get_folders_tree(self):
        try:
            resp = self._session.get("https://api.digiposte.fr/api/v3/folders", allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 200:
            print("get_folders_tree() HTTP error code:", resp.status_code)
            return "err"
        
        return resp.text
    
    def get_folder_content(self, folder_id):
        payload = {"locations": ["INBOX", "SAFE"], "folder_id": folder_id}
        
        try:
            resp = self._session.post("https://api.digiposte.fr/api/v3/documents/search?max_results=1000&sort=TITLE", json=payload, allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 200:
            print("get_folder_content() HTTP error code:", resp.status_code)
            return "err"
        
        return resp.text
    
    def get_file(self, file_id, dest_path):
        try:
            resp = self._session.get("https://api.digiposte.fr/api/v3/document/{}/content".format(file_id), allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 200:
            print("get_file() HTTP error code:", resp.status_code)
            return "err"
        
        try:
            with open(dest_path, 'wb') as f:
                f.write(resp.content)
        except Exception as e:
            print(e)
            return "err"
        
        return "OK"
    
    def create_folder(self, name, parent_id):
        payload = {"name": "{}".format(name), "favorite": False, "parent_id": "{}".format(parent_id)}
        
        try:
            resp = self._session.post("https://api.digiposte.fr/api/v3/folder", json=payload, allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 200:
            print("create_folder() HTTP error code:", resp.status_code)
            return "err"
        
        try:
            return resp.json()["id"]
        except requests.JSONDecodeError as e:
            print(e)
            return "err"
    
    def rename_object(self, object_id, new_name, is_file):
        if is_file:
            url = "https://api.digiposte.fr/api/v3/document/{}/rename/{}".format(object_id, new_name)
        else:
            url = "https://api.digiposte.fr/api/v3/folder/{}/rename/{}".format(object_id, new_name)
        
        try:
            resp = self._session.put(url, allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 200:
            print("rename_object() HTTP error code:", resp.status_code)
            return "err"
        
        return resp.text
    
    def delete_object(self, object_id, is_file):
        if is_file:
            payload = {"document_ids": ["{}".format(object_id)], "folder_ids": []}
        else:
            payload = {"document_ids": [], "folder_ids": ["{}".format(object_id)]}
        
        try:
            resp = self._session.post("https://api.digiposte.fr/api/v3/file/tree/trash", json=payload, allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 204:
            print("delete_object() HTTP error code:", resp.status_code)
            return "err"
        
        return resp.text
    
    def move_object(self, object_id, dest_folder_id, is_file):
        if is_file:
            payload = {"document_ids": ["{}".format(object_id)], "folder_ids": []}
        else:
            payload = {"document_ids": [], "folder_ids": ["{}".format(object_id)]}
        
        try:
            resp = self._session.put("https://api.digiposte.fr/api/v3/file/tree/move", params={"to": dest_folder_id}, json=payload, allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 204:
            print("move_object() HTTP error code:", resp.status_code)
            return "err"
        
        return resp.text
    
    def upload_file(self, dest_folder_id, src_file_path, name, size):
        try:
            f = open(src_file_path, 'rb')
        except Exception as e:
            print(e)
            return "err"
        
        if dest_folder_id is None:
            payload = {"archive_size": size, "archive": (name, f), "health_document": False, "title": name}
        else:
            payload = {"archive_size": size, "archive": (name, f), "health_document": False, "title": name, "folder_id": dest_folder_id}
        
        try:
            resp = self._session.post("https://api.digiposte.fr/api/v3/document", files=payload, allow_redirects=False)
        except requests.Timeout:
            return "err"
        except (requests.RequestException, requests.ConnectionError, requests.TooManyRedirects) as e:
            print(e)
            return "err"
        
        if resp.status_code != 200:
            print("upload_file() HTTP error code:", resp.status_code)
            return "err"
        
        f.close()
        try:
            return resp.json()["id"]
        except requests.JSONDecodeError as e:
            print(e)
            return "err"

if __name__ == "__main__":
    parser = argparse.ArgumentParser(prog="DigiposteAPI", description="Digiposte API communication")
    parser.add_argument("--server", nargs=2, metavar=("read_fd", "write_fd"), type=int, required=False, help="Spawn DigiposteAPI as a server with anonymous pipe passed as arguments")
    parser.add_argument("--token", nargs=1, metavar="token", required=False, help="Authentication token for Digiposte API. Not recommended")
    #parser.add_argument("--retry", nargs=1, type=int, required=False, default=3, help="Number of retries when API returns an error. Default to 3")
    #parser.add_argument("--cli", action='store_true', default=False, help="Prompts will be printed in terminal, otherwise use UI")
    
    subparsers = parser.add_subparsers(metavar="action", dest="action", required=False, help='Action to perform')
    parser_get_folders_tree = subparsers.add_parser('get_folders_tree', help='Get folders tree')
    
    parser_get_folder_content = subparsers.add_parser('get_folder_content', help='Get folder content')
    parser_get_folder_content.add_argument("folder_id", help="Folder ID to retrieve content from")
    
    parser_get_file = subparsers.add_parser('get_file', help='Get file')
    parser_get_file.add_argument("file_id", help="File ID to download")
    parser_get_file.add_argument("dest_path", help="Destination path for the downloaded file")
    
    parser_create_folder = subparsers.add_parser('create_folder', help='Create folder')
    parser_create_folder.add_argument("name", help="Name of the new folder")
    parser_create_folder.add_argument("parent_id", help="Folder ID of the parent folder")
    
    parser_rename_object = subparsers.add_parser('rename_object', help='Rename object')
    parser_rename_object.add_argument("--file", required=False, help="The object is a file")
    parser_rename_object.add_argument("object_id", help="Object ID to rename")
    parser_rename_object.add_argument("new_name", help="New name of the object")
    
    parser_delete_object = subparsers.add_parser('delete_object', help='Delete object')
    parser_delete_object.add_argument("--file", required=False, help="The object is a file")
    parser_delete_object.add_argument("object_id", help="Object ID to delete")
    
    parser_move_object = subparsers.add_parser('move_object', help='Move object')
    parser_move_object.add_argument("--file", required=False, help="The object is a file")
    parser_move_object.add_argument("object_id", help="Object ID to move")
    parser_move_object.add_argument("dest_folder_id", help="Destination folder ID")
    
    parser_upload_file = subparsers.add_parser('upload_file', help='Upload file')
    parser_upload_file.add_argument("dest_folder_id", help="Destination folder ID")
    parser_upload_file.add_argument("src_file_path", help="Path of the file to upload")
    parser_upload_file.add_argument("name", help="Name of the file to upload")
    parser_upload_file.add_argument("size", type=int, help="Size of the file to upload in bytes")
    
    args = parser.parse_args()
    
    dgp_api = DigiposteAPI(token=args.token)
    
    if args.server:
        read_f = os.fdopen(args.server[0], mode='rb')
        write_fd = args.server[1]
        
        os.write(write_fd, b"ready" + b'\0')
        
        buffer = read_f.readline()
        while buffer != b"":
            com = buffer[:-1].split(b'\0')
            if com[0] == b"get_folders_tree":
                tree = dgp_api.get_folders_tree()
                os.write(write_fd, tree.encode() + b'\0')
            
            elif com[0] == b"get_folder_content":
                content = dgp_api.get_folder_content(com[1].decode())
                os.write(write_fd, content.encode() + b'\0')
            
            elif com[0] == b"get_file":
                if dgp_api.get_file(com[1].decode(), com[2].decode()) == "err":
                    os.write(write_fd, b'err' + b'\0')
                else:
                    os.write(write_fd, b'OK' + b'\0')
            
            elif com[0] == b"create_folder":
                folder_id = dgp_api.create_folder(com[1].decode(), com[2].decode())
                os.write(write_fd, folder_id.encode() + b'\0')
            
            elif com[0] == b"rename_object":
                is_file = com[1] == b'1'
                if dgp_api.rename_object(com[2].decode(), com[3].decode(), is_file) == "err":
                    os.write(write_fd, b'err' + b'\0')
                else:
                    os.write(write_fd, b'OK' + b'\0')
            
            elif com[0] == b"delete_object":
                is_file = com[1] == b'1'
                if dgp_api.delete_object(com[2].decode(), is_file) == "err":
                    os.write(write_fd, b'err' + b'\0')
                else:
                    os.write(write_fd, b'OK' + b'\0')
            
            elif com[0] == b"move_object":
                is_file = com[1] == b'1'
                if com[3] == b'':
                    dest_folder_id = None
                else:
                    dest_folder_id = com[3].decode()
                
                if dgp_api.move_object(com[2].decode(), dest_folder_id, is_file) == "err":
                    os.write(write_fd, b'err' + b'\0')
                else:
                    os.write(write_fd, b'OK' + b'\0')
            
            elif com[0] == b"upload_file":
                if com[1] == b'':
                    dest_folder_id = None
                else:
                    dest_folder_id = com[1].decode()
                file_id = dgp_api.upload_file(dest_folder_id, com[2].decode(), com[3].decode(), com[4].decode())
                os.write(write_fd, file_id.encode() + b'\0')
            
            else:
                print("Unknown command", com[0], "with parameters", buffer[:-1].split(b'\0')[1:])
                os.write(write_fd, b'err' + b'\0')
            
            buffer = read_f.readline()
        
        print("Pipe closed by client. Exiting...")
        dgp_api.disconnect()
        read_f.close()
        os.close(write_fd)
        
    else:
        if args.action == "get_folders_tree":
            print(dgp_api.get_folders_tree())
        elif args.action == "get_folder_content":
            pass
        elif args.action == "get_file":
            pass
        elif args.action == "create_folder":
            pass
        elif args.action == "rename_object":
            pass
        elif args.action == "delete_object":
            pass
        elif args.action == "move_object":
            pass
        elif args.action == "upload_file":
            pass
        else:
            print("Error")
